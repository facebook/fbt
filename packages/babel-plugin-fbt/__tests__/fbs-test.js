/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 *
 * This file is shared between www and fbsource and www is the source of truth.
 * When you make change to this file on www, please make sure you test it on
 * fbsource and send a diff to update the files too so that the 2 versions are
 * kept in sync.
 *
 * Run the following command to sync the change from www to fbsource.
 *   js1 upgrade www-shared -p babel_plugin_fbt --local ~/www
 *
 * @emails oncall+internationalization
 * @flow
 * @format
 */
/*global describe, it, expect*/

jest.autoMockOff();

const {transform, withFbsRequireStatement} = require('../FbtTestUtil');

describe('Test declarative (jsx) <fbs> syntax translation', () => {
  it('should convert a simple string', () => {
    expect(
      transform(
        withFbsRequireStatement(
          `const fbsElem = <fbs desc='str_description'>a simple string</fbs>;`,
        ),
      ),
    ).toMatchSnapshot();
  });
  it('should convert a string with a parameter', () => {
    expect(
      transform(
        withFbsRequireStatement(`
          const fbsElem = <fbs desc='str_description'>
            a string with a
            <fbs:param name="param name">{parameter}</fbs:param>
          </fbs>;
        `),
      ),
    ).toMatchSnapshot();
  });
  it('should convert a common string', () => {
    expect(
      transform(
        withFbsRequireStatement(`
          const fbsCommonElem = <fbs common={true}>Post</fbs>;
        `),
      ),
    ).toMatchSnapshot();
  });
  it('should reject an <fbs> child element', () => {
    expect(() =>
      transform(
        withFbsRequireStatement(`
          const fbsElem = <fbs desc='str_description'>
            a simple string
            <fbs>nested</fbs>
          </fbs>;
        `),
      ),
    ).toThrow(`Don't put <fbs> directly within <fbs>.`);
  });
  it('should reject an <fbt> child element', () => {
    expect(() =>
      transform(
        withFbsRequireStatement(`
          const fbsElem = <fbs desc='str_description'>
            a simple string
            <fbt>nested</fbt>
          </fbs>;
        `),
      ),
    ).toThrow(`Don't put <fbt> directly within <fbs>.`);
  });
  it('should reject an <fbt:param> child element', () => {
    expect(() =>
      transform(
        withFbsRequireStatement(`
          const fbsElem = <fbs desc='str_description'>
            a simple string
            <fbt:param name="param name">{parameter}</fbt:param>
          </fbs>;
        `),
      ),
    ).toThrow(`Don't mix <fbt> and <fbs> JSX namespaces.`);
  });
});

describe('Test functional fbs() syntax translation', () => {
  it('should convert a simple string', () => {
    expect(
      transform(
        withFbsRequireStatement(
          `const fbsCall = fbs('a simple string', 'str_description');`,
        ),
      ),
    ).toMatchSnapshot();
  });
  it('should convert a string with a gender parameter', () => {
    expect(
      transform(
        withFbsRequireStatement(`
          const fbsCall = fbs(
            'a string with a ' + fbs.param('param name', parameter, {gender: 'male'}),
            'str_description'
          );
        `),
      ),
    ).toMatchSnapshot();
  });
  it('should convert a common string', () => {
    expect(
      transform(
        withFbsRequireStatement(`const fbsCommonCall = fbs.c('Post');`),
      ),
    ).toMatchSnapshot();
  });
  it('should reject an fbt parameter', () => {
    expect(() =>
      transform(
        withFbsRequireStatement(`
          const fbsCall = fbs(
            'a string with a ' + fbt.param('param name', parameter, {gender: 'male'}),
            'str_description'
          );
        `),
      ),
    ).toThrow(
      `fbs only accepts plain strings with params wrapped in fbs.param.`,
    );
  });

  it('should throw when using fbs() and the fbs variable is not bound', () => {
    expect(() =>
      transform(
        `const fbsCall = fbs(
          'basic',
          'str_description'
        );`,
      ),
    ).toThrow(`fbs is not bound. Did you forget to require('fbs')?`);
  });

  it('should throw when using <fbs> and the fbs variable is not bound', () => {
    expect(() =>
      transform(`const fbsCall = <fbs desc="str_description">basic</fbs>;`),
    ).toThrow(`fbs is not bound. Did you forget to require('fbs')?`);
  });
});
