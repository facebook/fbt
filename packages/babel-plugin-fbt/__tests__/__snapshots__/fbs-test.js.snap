// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test declarative (jsx) <fbs> syntax translation should convert a common string 1`] = `
"const fbs = require(\\"fbs\\");

const fbsCommonElem = fbs._(\\"__FBT__{\\\\\\"type\\\\\\":\\\\\\"text\\\\\\",\\\\\\"jsfbt\\\\\\":\\\\\\"Post\\\\\\",\\\\\\"desc\\\\\\":\\\\\\"Button to post a comment\\\\\\",\\\\\\"project\\\\\\":\\\\\\"\\\\\\"}__FBT__\\");"
`;

exports[`Test declarative (jsx) <fbs> syntax translation should convert a simple string 1`] = `
"const fbs = require(\\"fbs\\");

const fbsElem = fbs._(\\"__FBT__{\\\\\\"type\\\\\\":\\\\\\"text\\\\\\",\\\\\\"jsfbt\\\\\\":\\\\\\"a simple string\\\\\\",\\\\\\"desc\\\\\\":\\\\\\"str_description\\\\\\",\\\\\\"project\\\\\\":\\\\\\"\\\\\\"}__FBT__\\");"
`;

exports[`Test declarative (jsx) <fbs> syntax translation should convert a string with a parameter 1`] = `
"const fbs = require(\\"fbs\\");

const fbsElem = fbs._(\\"__FBT__{\\\\\\"type\\\\\\":\\\\\\"text\\\\\\",\\\\\\"jsfbt\\\\\\":\\\\\\"a string with a {param name}\\\\\\",\\\\\\"desc\\\\\\":\\\\\\"str_description\\\\\\",\\\\\\"project\\\\\\":\\\\\\"\\\\\\"}__FBT__\\", [fbs._param(\\"param name\\", parameter)]);"
`;

exports[`Test functional fbs() syntax translation should convert a common string 1`] = `
"const fbs = require(\\"fbs\\");

const fbsCommonCall = fbs._(\\"__FBT__{\\\\\\"type\\\\\\":\\\\\\"text\\\\\\",\\\\\\"jsfbt\\\\\\":\\\\\\"Post\\\\\\",\\\\\\"desc\\\\\\":\\\\\\"Button to post a comment\\\\\\",\\\\\\"project\\\\\\":\\\\\\"\\\\\\"}__FBT__\\");"
`;

exports[`Test functional fbs() syntax translation should convert a simple string 1`] = `
"const fbs = require(\\"fbs\\");

const fbsCall = fbs._(\\"__FBT__{\\\\\\"type\\\\\\":\\\\\\"text\\\\\\",\\\\\\"jsfbt\\\\\\":\\\\\\"a simple string\\\\\\",\\\\\\"desc\\\\\\":\\\\\\"str_description\\\\\\",\\\\\\"project\\\\\\":\\\\\\"\\\\\\"}__FBT__\\");"
`;

exports[`Test functional fbs() syntax translation should convert a string with a gender parameter 1`] = `
"const fbs = require(\\"fbs\\");

const fbsCall = fbs._(\\"__FBT__{\\\\\\"type\\\\\\":\\\\\\"table\\\\\\",\\\\\\"jsfbt\\\\\\":{\\\\\\"t\\\\\\":{\\\\\\"*\\\\\\":\\\\\\"a string with a {param name}\\\\\\"},\\\\\\"m\\\\\\":[{\\\\\\"token\\\\\\":\\\\\\"param name\\\\\\",\\\\\\"type\\\\\\":1}]},\\\\\\"desc\\\\\\":\\\\\\"str_description\\\\\\",\\\\\\"project\\\\\\":\\\\\\"\\\\\\"}__FBT__\\", [fbs._param('param name', parameter, [1, 'male'])]);"
`;
